File: esbmc.cpp/validador_categoria/teste.cpp
#include <cctype>
#include <cstdlib>
#include <cstring>

#define MAX_LEN 31 

using namespace std;

bool validate_category(const char category[MAX_LEN], int len) {
    const char specials[] = "!@#$%^&*()_+=-[]{}|;:,.<>?/~`";

    __ESBMC_assert(len >= 3 && len <= 30, "A categoria deve ter entre 3 e 30 caracteres.");

    __ESBMC_assert(isalpha(category[0]), "A categoria deve começar com uma letra.");

    for (int i = 0; i < len - 1; i++) {
        __ESBMC_assert(category[i] != category[i + 1], "A categoria não deve conter caracteres repetidos consecutivos.");
    }

    for (int i = 0; i < len; i++) {
        __ESBMC_assert(category[i] != ' ', "A categoria não deve conter espaços.");
    }

    for (int i = 0; i < len; i++) {
        for (int j = 0; specials[j] != '\0'; j++) {
            __ESBMC_assert(category[i] != specials[j], "A categoria não deve conter caracteres especiais.");
        }
    }

    char last = category[len - 1];
    for (int j = 0; specials[j] != '\0'; j++) {
        __ESBMC_assert(last != specials[j], "A categoria não pode terminar com um número ou caractere especial.");
    }
    __ESBMC_assert(!isdigit(last), "A categoria não pode terminar com um número ou caractere especial.");

    for (int i = 0; i < len; i++) {
        __ESBMC_assert(isalpha(category[i]), "A categoria deve conter apenas letras (sem números).");
    }

    return true;
}

int main() {
    char test[MAX_LEN];
    int len;

    __ESBMC_assume(len >= 3 && len <= 30);

    for (int i = 0; i < len; i++) {
        __ESBMC_assume(test[i] >= 32 && test[i] <= 126);
    }

    for (int i = len; i < MAX_LEN; i++) {
        test[i] = '\0';
    }

    validate_category(test, len);

    return 0;
}

File: esbmc.py/Testes02/validador_categoria/teste.py
MAX_LEN: int = 31

def is_letter(c: str) -> bool:
    return ('a' <= c <= 'z') or ('A' <= c <= 'Z')

def validate_category(category: str, length: int) -> bool:
    specials: str = "!@#$%^&*()_+=-[]{}|;:,.<>?/~`"

    assert 3 <= length <= 30, "A categoria deve ter entre 3 e 30 caracteres."

    assert is_letter(category[0]), "A categoria deve começar com uma letra."

    i: int = 0
    while i < length - 1:
        assert category[i] != category[i + 1], \
            "A categoria não deve conter caracteres repetidos consecutivos."
        i += 1

    i = 0
    while i < length:
        c: str = category[i]
        assert c != ' ', "A categoria não deve conter espaços."
        assert is_letter(c), "A categoria deve conter apenas letras (sem números)."
        
        j: int = 0
        while j < len(specials):
            assert c != specials[j], "A categoria não deve conter caracteres especiais."
            j += 1
        i += 1

    last: str = category[length - 1]
    assert not ('0' <= last <= '9') and last not in specials, \
        "A categoria não pode terminar com um número ou caractere especial."

    return True

def main() -> None:
    category: str = input()
    length: int = len(category)

    assert 3 <= length <= 30

    validate_category(category, length)

main()